sdk/src/main/scala/com/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala

```diff
@@ -0,0 +1,524 @@
+      btrFee,
+      ftMinAmount,
+      utxoMerkleTreeRoot,
+      signersPublicKeyWithSignatures)
+  }
+
+  private def lastMainchainBlockCumulativeCommTreeHashForWithdrawalEpochNumber(history: SidechainHistory, withdrawalEpochNumber: Int): Array[Byte] = {
+    val mcBlockHash = withdrawalEpochNumber match {
+      case -1 => params.parentHashOfGenesisMainchainBlock
+      case _ => {
+        val mcHeight = params.mainchainCreationBlockHeight + (withdrawalEpochNumber + 1) * params.withdrawalEpochLength - 1
+        history.getMainchainBlockReferenceInfoByMainchainBlockHeight(mcHeight).asScala.map(_.getMainchainHeaderHash).getOrElse(throw new IllegalStateException("Information for Mc is missed"))
+      }
+    }
+    log.debug(s"Last MC block hash for withdrawal epoch number $withdrawalEpochNumber is ${
+      BytesUtils.toHexString(mcBlockHash)
+    }")
+
+    val headerInfo = history.mainchainHeaderInfoByHash(mcBlockHash).getOrElse(throw new IllegalStateException("Missed MC Cumulative Hash"))
+
+    headerInfo.cumulativeCommTreeHash
+  }
+
+  private[certificatesubmitter] def generateProof(dataForProofGeneration: DataForProofGeneration): com.horizen.utils.Pair[Array[Byte], java.lang.Long] = {
+    val (signersPublicKeysBytes: Seq[Array[Byte]], signaturesBytes: Seq[Optional[Array[Byte]]]) =
+      dataForProofGeneration.schnorrKeyPairs.map {
+        case (proposition, proof) => (proposition.bytes(), proof.map(_.bytes()).asJava)
+      }.unzip
+
+    log.info(s"Start generating proof with parameters: dataForProofGeneration = ${
+      dataForProofGeneration
+    }, " +
+      s"signersThreshold = ${
+        params.signersThreshold
+      }. " +
+      s"It can take a while.")
+
+    //create and return proof with quality
+    CryptoLibProvider.sigProofThresholdCircuitFunctions.createProof(
+      dataForProofGeneration.withdrawalRequests.asJava,
+      dataForProofGeneration.sidechainId,
+      dataForProofGeneration.referencedEpochNumber,
+      dataForProofGeneration.endEpochCumCommTreeHash,
+      dataForProofGeneration.btrFee,
+      dataForProofGeneration.ftMinAmount,
+      dataForProofGeneration.utxoMerkleTreeRoot,
+      signaturesBytes.asJava,
+      signersPublicKeysBytes.asJava,
+      params.signersThreshold,
+      provingFileAbsolutePath,
+      true,
+      true)
+  }
+
+  def submitterStatus: Receive = {
+    case EnableSubmitter =>
+      submitterEnabled = true
+    case DisableSubmitter =>
+      submitterEnabled = false
+    case IsSubmitterEnabled =>
+      sender() ! submitterEnabled
+  }
+
+  def signerStatus: Receive = {
+    case EnableCertificateSigner =>
+      // Next signing attempt will be at the beginning of the next submission window.
+      certificateSigningEnabled = true
+    case DisableCertificateSigner =>
+      certificateSigningEnabled = false
+    case IsCertificateSigningEnabled =>
+      sender() ! certificateSigningEnabled
+  }
+}
+
+object AbstractCertificateSubmitter {
```

Comment:
all the events, interfaces and so on must be declared here and reused in CertificateSubmitter and NonCeasingCertSubmitter.